contributors/design-proposals/api-paging.md

```diff
@@ -0,0 +1,148 @@
+# Reduce the amount of client and server memory allocated during large API lists
+
+On large clusters, performing API queries that return all of the objects of a given resource type (GET /api/v1/pods, GET /api/v1/secrets) can lead to significant variations in peak memory use on the server and contribute substantially to long tail request latency.
+
+When loading very large sets of objects -- some clusters are now reaching 100k pods or equivalent numbers of supporting resources -- the system must:
+
+* Construct the full range description in etcd in memory and serialize it as protobuf in the client
+  * Some clusters have reported over 500MB being stored in a single object type
+  * This data is read from the underlying datastore and converted to a protobuf response
+  * Large reads to etcd can block writes to the same range (https://github.com/coreos/etcd/issues/7719)
+* The data from etcd has to be transferred to the apiserver in one large chunk
+* The `kube-apiserver` also has to deserialize that response into a single object, and then re-serialize it back to the client
+  * Much of the decoded etcd memory is copied into the struct used to serialize to the client
+* A client like `kubectl get` will then decode the response from JSON or protobuf
+  * A client with a slow connection may not be able to receive the entire response body within the default 60s timeout
+  * The recently introduced client compression feature can assist
+  * The large response will also be loaded entirely into memory
+
+The standard solution for reducing the impact of large reads is to allow them to be broken into smaller reads via a technique commonly referred to as paging. By efficiently splitting large list ranges from clients to etcd into many smaller list ranges, we can reduce the maximum memory allocation on etcd and the apiserver.
+
+Our primary consistent store etcd3 offers support for efficient paging with minimal overhead, and mechanisms exist for other potential future stores such as SQL databases or Consul to also implement a simple form of paging.
+
+Relevant issues:
+
+* https://github.com/kubernetes/kubernetes/issues/2349
+
+## Proposed change:
+
+Expose a simple paging mechanism to allow large API responses to be broken into consistent partial responses. Clients would indicate a tolerance for paging (opt-in) by specifying a desired maximum number of results to return in a `LIST` call. The server would return up to that amount of objects, and if more exist it would return a `continue` parameter that the client could pass to receive the next set of results.  The server would be allowed to ignore the limit if it does not implement limiting (backward compatible), but it is not allowed to support limiting without supporting a way to continue the query past the limit (may not implement `limit` without `continue`).
```

Comment:
> is there a life-time of such a continue parameter? 

It's up to the backing store to have the historical data available.  So for our default config is 5m for etcd3.

> If the client asked for just 5 because it can't handle more, I'm not sure the server should be allowed to ignore the request and flood them with a massive response. I'd prefer an error that tells the client that the server doesn't support paging and then let the client decide what to do next.

That is not backwards compatible - old servers wouldn't return an error.  We generally err on not introducing new incompatibilities.  Paging is a way to get less results in a chunk, not a way to slowly browse through the data store.  So naive web clients that want to display a chunk of results at a time needs something more involved than simple continuations and would be an additional set of parameters.